server/src/main.rs

```diff
@@ -0,0 +1,156 @@
+extern crate hyper;
+
+use hyper::{service, Request, Response, Body, Server, StatusCode};
+use hyper::header::{HeaderMap, HeaderName, HeaderValue};
+use futures::{future::{self, Either}, Future, Stream};
+use serde_json::json;
+use lazy_static::lazy_static;
+use ed25519_dalek::{PublicKey, Signature};
+use std::{env, fs};
+use hpos_state_core::state::State;
+use base64::decode_config;
+
+lazy_static! {
+    static ref X_HPOS_ADMIN_SIGNATURE: HeaderName = HeaderName::from_lowercase(b"x-hpos-admin-signature").unwrap();
+    static ref X_ORIGINAL_URI: HeaderName = HeaderName::from_lowercase(b"x-original-uri").unwrap();
+    static ref HP_PUBLIC_KEY: PublicKey = read_hp_pubkey();
+}
+
+// Create response based on the request parameters
+fn create_response(req: Request<Body>) -> impl Future<Item = Response<Body>, Error = hyper::Error> {
+    let (parts, body) = req.into_parts();
+
+    match parts.uri.path() {
+        "/" => {
+            let entire_body = body.concat2();
+            let res = entire_body.map( |body| {
+                // Extract X-Original-URI header value, panic for no header
+                let req_uri = match parts.headers.get(&*X_ORIGINAL_URI) {
+                    Some(s) => s.to_str().unwrap(),
+                    None => panic!("Request does not contain \"X-Original-URI\" header."),
+                };
+                let body_string = String::from_utf8(body.to_vec()).expect("Found invalid UTF-8");
+                let payload = create_payload(parts.method.to_string(), req_uri.to_string(), body_string);
+                let is_verified = verify_request(payload, parts.headers, &HP_PUBLIC_KEY);
+                respond_success(is_verified)
+            });
+
+            Either::A(res)
+        }
+        _ => {
+            let res = future::ok(respond_success(false));
+            Either::B(res)
+        }
+    }
+}
+
+fn create_payload (method: String, uri: String, body_string: String) -> String {
+    let d = json!({
```

Comment:
This should be a struct of its own rather than set inline.

Reply:
done